safehttp/incoming_request_test.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package safehttp_test
+
+import (
+	"context"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-safeweb/safehttp"
+	"net/http/httptest"
+	"testing"
+)
+
+type pizza struct {
+	val string
+}
+
+type pizzaKey string
+
+func TestRequestSetValidContextWithValue(t *testing.T) {
+	tests := []struct {
+		name    string
+		key     pizzaKey
+		wantVal *pizza
+		wantOk  bool
+	}{
+		{
+			name:    "Value set for key",
+			key:     pizzaKey("1234"),
+			wantOk:  true,
+			wantVal: &pizza{val: "margeritta"},
+		},
+		{
+			name:    "Value not set for key",
+			key:     pizzaKey("5678"),
+			wantOk:  false,
+			wantVal: nil,
+		},
+	}
+	for _, test := range tests {
+		req := httptest.NewRequest("GET", "/", nil)
+		ir := safehttp.NewIncomingRequest(req)
+		ctx := context.WithValue(ir.Context(), pizzaKey("1234"), &pizza{val: "margeritta"})
+		ir.SetContext(ctx)
+
+		got, ok := ir.Context().Value(test.key).(*pizza)
+		if ok != test.wantOk {
+			t.Errorf("type match: got %v, want %v", ok, test.wantOk)
+		}
+		if diff := cmp.Diff(test.wantVal, got, cmp.AllowUnexported(pizza{})); diff != "" {
+			t.Errorf("ir.Context().Value(k): mismatch (-want +got): \n%s", diff)
+		}
+	}
+}
+
+func TestRequestSetNilContext(t *testing.T) {
+	defer func() {
+		if r := recover(); r == nil {
+			t.Errorf(`ir.SetContext(nil): expected panic`)
+		}
+	}()
```

Comment:
Move this `defer` to right before `ir.SetContext(nil)`, otherwise this could potentially recover panics from `httptest.NewRequest` and `safehttp.NewIncominRequest`